src/python/bot/tasks/fuzz_task.py

```diff
@@ -1254,319 +1258,391 @@ def get_strategy_distribution_from_ndb():
+        data_types.Fuzzer.name == self.fuzzer_name).get()
+    if not fuzzer or not setup.update_fuzzer_and_data_bundles(self.fuzzer_name):
+      _track_fuzzer_run_result(self.fuzzer_name, 0, 0,
+                               FuzzErrorCode.FUZZER_SETUP_FAILED)
+      logs.log_error('Unable to setup fuzzer %s.' % self.fuzzer_name)
+
+      # Artifical sleep to slow down continuous failed fuzzer runs if the bot is
+      # using command override for task execution.
+      time.sleep(failure_wait_interval)
+      return
 
-    process_handler.close_queue(temp_queue)
+    # Set up a custom or regular build based on revision. By default, fuzzing
+    # is done on trunk build (using revision=None). Otherwise, a job definition
+    # can provide a revision to use via |APP_REVISION|.
+    if (build_manager.setup_build(environment.get_value('APP_REVISION')) and
+        environment.get_value('DATAFLOW_BUILD_BUCKET_PATH')):
+      # Some fuzzing jobs may use auxiliary builds, such as DFSan instrumented
+      # builds accompanying libFuzzer builds to enable DFT-based fuzzing.
+      build_manager.setup_trunk_build(['DATAFLOW_BUILD_BUCKET_PATH'],
+                                      'DATAFLOW')
+
+    # Check if we have an application path. If not, our build failed
+    # to setup correctly.
+    app_path = environment.get_value('APP_PATH')
+    if not app_path:
+      _track_fuzzer_run_result(self.fuzzer_name, 0, 0,
+                               FuzzErrorCode.BUILD_SETUP_FAILED)
+      return
 
-    logs.log('Upto %d' % test_number)
+    # Check if we have a bad build, i.e. one that crashes on startup.
+    # If yes, bail out.
+    logs.log('Checking for bad build.')
+    crash_revision = environment.get_value('APP_REVISION')
+    is_bad_build = testcase_manager.check_for_bad_build(self.job_type,
+                                                        crash_revision)
+    _track_build_run_result(self.job_type, crash_revision, is_bad_build)
+    if is_bad_build:
+      return
 
-    if thread_error_occurred:
-      break
+    # Data bundle directories can also have testcases which are kept in-place
+    # because of dependencies.
+    self.data_directory = setup.get_data_bundle_directory(self.fuzzer_name)
+    if not self.data_directory:
+      _track_fuzzer_run_result(self.fuzzer_name, 0, 0,
+                               FuzzErrorCode.DATA_BUNDLE_SETUP_FAILED)
+      logs.log_error(
+          'Unable to setup data bundle %s.' % fuzzer.data_bundle_name)
+      return
 
-  # Pull testcase directory to host. The testcase file contents could have been
-  # changed (by e.g. libFuzzer) and stats files could have been written.
-  if environment.is_trusted_host():
-    from bot.untrusted_runner import file_host
-    file_host.pull_testcases_from_worker()
+    engine_impl = engine.get(fuzzer.name)
+    if engine_impl:
+      crashes = self.do_engine_fuzzing(engine_impl)
+      # TODO(ochang): fill these.
+      testcase_file_paths = []
```

Comment:
Dont understand this, how can we proceed without these ? So this "if engine_impl" branch is dead currently, probably a FIXME or TODO to indicate that.

Reply:
This is just a placeholder for future implementation. Added a TODO.